src/common/tensor_inspector.h

```diff
@@ -0,0 +1,733 @@
+  inline void check_value_helper(std::vector<std::vector<int>>* ret,
+      const std::function<bool(DType)>& checker, bool interactive, std::string tag) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      return TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .check_value_helper<DType>(ret, checker, interactive, tag);
+    }
+#endif  // MXNET_USE_CUDA
+    size_t count = 0;
+    std::stringstream ss;
+    ss << "[";
+    bool first_pass = true;
+    for (size_t i = 0; i < tb_.shape_.Size(); ++i) {
+      if (checker(tb_.dptr<DType>()[i])) {
+        count += 1;
+        if (!first_pass) {
+          ss  << ", ";
+        }
+        first_pass = false;
+        std::vector<int> coords = index_to_coordinates(i);
+        ss << "(" << coords[0];
+        for (unsigned int i = 1; i < coords.size(); ++i) {
+          ss << ", " << coords[i];
+        }
+        ss << ")";
+        ret->push_back(coords);
+      }
+    }
+    ss << "]" << std::endl;
+    if (interactive) {
+      std::lock_guard<std::mutex> lock(InspectorManager::get()->mutex_);
+       InspectorManager::get()->check_value_tag_counter_[tag] += 1;
+      while (!InspectorManager::get()->check_value_skip_all_) {
+        std::cout << "----------Value Check----------" << std::endl;
+        if (tag != "") {
+          std::cout << "Tag: " << tag << "  Visit: " <<
+              InspectorManager::get()->check_value_tag_counter_[tag] <<  std::endl;
+        }
+        std::cout << count << " value(s) found." << std::endl;
+        std::cout << "\"p\" to print the coordinates, \"b\" to break, \"s\" to skip all: ";
+        std::string str;
+        std::cin >> str;
+        if (str == "b") {
+          break;
+        } else if (str == "p") {
+          std::cout << ss.str() << std::endl;
+        } else if (str == "s") {
+          InspectorManager::get()->check_value_skip_all_ = true;
+        }
+      }
+    }
+  }
+
+  /*!
+   * \brief infer the python type, given the c++ type
+   * \tparam ti the type info 
+   */
+  inline char infer_type(const std::type_info& ti) {
+    if (ti == typeid(float)) return 'f';
+    else if (ti == typeid(double)) return 'f';
+    else if (ti == typeid(mshadow::half::half_t) ) return 'f';
+    else if (ti == typeid(uint8_t)) return 'u';
+    else if (ti == typeid(int32_t)) return 'i';
+    else if (ti == typeid(int64_t)) return 'i';
+    else
+      return '?';
+  }
+
+  /*!
+   * \brief check if the host machine is big or small endian
+   */
+  inline char endian_test() {
+    int x = 1;
+    return (reinterpret_cast<char*>(&x)[0]) ? '<' : '>';
+  }
+
+  /*!
+   * \brief dump the value of the tensor to a file with name "[tag]_[visit count].npy" in npy format
+   * the dump file follows npy 1.0 stantand
+   * \tparam DType the data type
+   * \param tag the name given to this call
+   */
+  template<typename DType MSHADOW_DEFAULT_DTYPE>
+  inline void dump_value_helper(const std::string& tag) {
```

Comment:
Can we factor out formatting and IO in separate fns? One function does the formatting and another one does the IO, this is more testable and modular. Also allows to reuse the formatting function....

Reply:
Yes!

Reply:
separated them